### ISSUE TITLE
for-yield in new syntax behaves differently depending on where it stands

### ISSUE THREAD

**Issue by buzden** [lampepfl/dotty#7307]
## first rewrite

Consider we've had the following simple `for`-`yield` code:
```Scala
def f(ts: List[Int]): F[List[Whatever]] = ts.traverse { t => for {
  r <- f(t)
} yield whatever(r)
}
```

`yield` in new-syntax can go without `<outdent>` when `for` is placed inside curly braces, e.g. inside an old-syntaxed lambda. So we can rewrite the original example like this:

```Scala
def f(ts: List[Int]): F[List[Whatever]] = ts.traverse { t => for
  r <- f(t)
  yield whatever(r)
}
```

This looks nice when rewriting old syntax code to the new one, since all left in their places, just several curly braces were gone.

## expectation

I was expected that this behaviour of `yield` work for every new-syntax `for`-`yield`. But when I was rewrite code like this

```scala
def g[F[_]: Monad](r: F[Int]): F[Int] = for {
  v <- r
} yield v
```

to this

```scala
def g[F[_]: Monad](r: F[Int]): F[Int] = for 
  v <- r
  yield v
```

dotty complains

```
[error] -- [E040] Syntax Error: .../src/main/scala/Main.scala:25:2
[error] 25 |  yield v
[error]    |  ^^^^^
[error]    |  unindent expected, but 'yield' found

```

So, is this behaviour what was planned with the new control syntax? Maybe, unindent should not be required before `yield` when no original identation was after `for`?

**Comment by aappddeevv** [comment]
Doesn't your yield have to indent with the `def` in the above example to be consistent with the indent rules?

**Comment by buzden** [comment]
> Doesn't your yield have to indent with the `def` in the above example to be consistent with the indent rules?

I'd say, _if_ `yield` is indented with the `def`, everything compiles, yes. But
- this requirements seems (to me, at least) rather strange and bad-looking;
- it is inconsistent with the case where the entire `for`-`yield` is in the block (as I shown in the first example).

**Comment by aappddeevv** [comment]
I hear you on that. If indent syntax leads to weird things, I just newline the "for" and the indent looks nice off of that. Perhaps @odersky has good ideas of what to do though.

**Comment by buzden** [comment]
> I just newline the "for" and the indent looks nice off of that.

I'm afraid of going bikeshedding, but this seems to be as an extra line of meaningless code (if to put `for` on the newline alone) or irregularity of the first expression under `for` with the others (if to put `for` and the first expression one after another).

Moreover, as far as I know, `= for` and then newline is really common in monadic code.

**Comment by odersky** [comment]
It's actually a more general problem. What about this:
```
def f() = try
  body
  catch { ... }
```
Should that also work? Or should we insist to put the `catch` to the left, as in
```
def f() = try
  body
catch { ...}
```
I'd have a tendency to do the latter, and also to require
```scala
def f() =for
  a <- as
yield ...
```

**Comment by buzden** [comment]
> I'd have a tendency to do the latter, and also to require

Oh, that's a pity.

I realized that I felt `yield` in `for` wrong for the entire life! I thought of it similar to the last `return` in the `do` notation block in Haskell, i.e. like an operation _inside_ the `for`. But, as far as I see, you feel `yield` as a separate operation _after_ the `for`-expression. That's probably why you feel requiring `yield` outdented better.

I can understand this requirement in imperative constructs since, for example, bodies of the `try` and `catch` are not homogeneous and behave very differently. That's why it's important to visually see difference in indent between `try`'s body and `catch`es. The same even for `for`-`do` because one section is for setting the space of the iteration and another one is for action inside that space.

But `yield` seemed to be another story since (depending on the type) it can be perform not only after the operations in the `for` body but rather together with them. Well, `yield` is just a `map`, like those between `for` and `yield` are just `flatMap`s. They are of the same nature, unlike imperative constructs.

Well, I (and other people that are thinking like me, if some) will get accustomed to this requirement some day, if needed. But I'd like to ask why not to make outdent before `yield` optional? Since directly nested `for`-`yield`s do not make sense (because they always can be flattened to a single `for`), it seems to not to introduce any ambiguity or problem. Moreover, my first example shows that even compiler technically can work without this requirement.

**Comment by odersky** [comment]
> But I'd like to ask why not to make outdent before yield optional? Since directly nested for-yields do not make sense (because they always can be flattened to a single for), it seems to not to introduce any ambiguity or problem. 

It's possible with a hack but difficult to do it in a principled way. Principled means: We need a grammar that the parser can follow. I don't see a good way to do that. All attempts to fix the grammar to allow this look rather horrible.

**Comment by odersky** [comment]
I am closing this now, but am ready to re-open if someone shows my a nice grammar for this.
